kvclient: handle speculative descriptors in the range cache better

andreimatei opened the pull request:
kvclient: handle speculative descriptors in the range cache better …

This patch fixes an assertion in the range cache that would sometimes
trigger, thinking that it has found inconsistent descriptors - two
overlapping descriptors with the same generation, but otherwise
different. This shouldn't happen, at least in the opinion of the
assertion. But it can happen in case one of the two descriptors is
"speculative": an eviction token for the range cache can contain a
replacement descriptor which comes from an intent. That intent might
never commit, in which case the respective generation might eventually
be assigned to some other, different descriptor.

This patch improves the cache's handling of speculative descriptors, and
also speculative leases, by handling them more explicitly. Such
descriptors are now inserted into the cache with Generation=0, and our
descriptor comparison code learns to deal explicitly with them.

Fixes #50995

Release note: None

andreimatei committed:
kvclient: handle speculative descriptors in the range cache better

This patch fixes an assertion in the range cache that would sometimes
trigger, thinking that it has found inconsistent descriptors - two
overlapping descriptors with the same generation, but otherwise
different. This shouldn't happen, at least in the opinion of the
assertion. But it can happen in case one of the two descriptors is
"speculative": an eviction token for the range cache can contain a
replacement descriptor which comes from an intent. That intent might
never commit, in which case the respective generation might eventually
be assigned to some other, different descriptor.

The interface presented by the cache to packages outside KV
(kvbase.RangeCacheEntry) is improved; the implementation of
RangeCacheEntry is moved to the kv package, and an interface is left
behind that talks about speculative data.

This patch improves the cache's handling of speculative descriptors, and
also speculative leases, by handling them more explicitly. Such
descriptors are now inserted into the cache with Generation=0, and our
descriptor comparison code learns to deal explicitly with them.

Fixes #50995

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51395)
<!-- Reviewable:end -->

nvb reviewed (commented):
I don't fully understand the motivation for this change. Before making the change, I'd like to better understand how things will work with and without it. Specifically, what will happen to a RangeCache that contains a descriptor coming from an intent in the following three cases:
1. the split completes successfully and resolves the intent
2. the split fails and rolls back the intent
3. the split stalls while continuing to heartbeat its txn record

In all three cases, we want to minimize the disruption that the RangeCache will have on foreground traffic trying to access the data on the splitting range.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MCENowQ9RaPXhC1y3Hp:b-iutyhf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
In all 3 cases, the intent descriptor will be put in the cache upon eviction of the original (modulo the problem that said eviction needs to happen specifically through the `EvictionToken` coming from the original insertion of the pre-split descriptor; evictions resulting from other cache operations won't do).
This patch does not change whether the speculative descriptor makes it into the cache or not. It just changes how vulnerable it is once it made it into the cache. Before, it would make it into the cache with its speculative `Generation`. Which is bad in case 2 (intent getting rolled back) because then the future real descriptor with that `Generation` will have trouble getting into the cache because it won't be `NewerThan` the speculative one. Most likely what will happen is that the insertion of the future descriptor will panic because it will detect a non-copacetic desc with the same generation: https://github.com/cockroachdb/cockroach/blob/06cd1cbadee9cbbfb16d56aa1ed3f27568610702/pkg/kv/kvbase/range_cache.go#L122

This patch fixes this by inserting the post-split descriptor with a `Generation` of 0, which makes the respective cache entry look older than any would-be replacement - in particular than the replacement with the same `Generation` as the rolled-back intent, or even the original descriptor (with a lower generation) if it turns out that we have wrongly evicted it.
This works for the split case, but not for the merge case. For merges, let's say we had [a,b) being merged with [b,c). Once we evict [a,b) we'll try to insert the speculative [a,c). Inserting [a,c) with a generation of 0 will fail if we have [b,c) in the cache. This seems funky, but the current behavior (evicting [b,c)) also seems funky to me. It's one thing to replace empty space in the cache with a speculative descriptor, it's another thing to replace an entry like [b,c) which might be fine for all we know. But I could be convinced otherwise, and we could fix it by finding another way of tracking the "speculative" character of [a,c).

If you buy this, I can update the commit msg.
But I'm interested in discussing the possibility of getting out of this speculation game, just because it seems complex to me and I'm scared of it. In what circumstances does the current speculation actually help? It only helps nodes that read meta2 *during the split/merge txn*. Which should be ~nobody, right? I think a better way to minimize the impact of splits/merges is to gossip the descriptor updates (as @knz and I keep talking about doing). If we do that gossiping, then I guess we have an analogous question about whether we should gossip speculatively before the split commit, or after the commit. But I'd just start with after and take it from there.
On the other hand, I was thinking that, perhaps, the point of the current speculative descriptor is not really to help with minimizing the impact of splits on traffic; instead maybe it's required for splits to work in the first place - I kinda remember something about these transactions relying on these intents in a subtle way (perhaps for cleaning up the intent itself because, after the commit, only the intent has the right descriptor in it?). Do you know anything about this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MCEVEQQ0qAXyvUAb15f:b-24jvc9)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
> On the other hand, I was thinking that, perhaps, the point of the current speculative descriptor is not really to help with minimizing the impact of splits on traffic; instead maybe it's required for splits to work in the first place - I kinda remember something about these transactions relying on these intents in a subtle way (perhaps for cleaning up the intent itself because, after the commit, only the intent has the right descriptor in it?). Do you know anything about this?

I don't recall the exact details, but it is the case that the speculative descriptors are required for splits to work correctly. I think you're right that it has to do with resolving the range-local intent on the RHS of the split after the split has committed. I wonder if any tests fail if you stop returning intents from READ_UNCOMMITTED scans.

I do support being more clear about when range descriptor information is speculative and when it's not. Being clear about that distinction seems to open up more room for being smart about caching and generally performing incremental cache updates like you're doing in https://github.com/cockroachdb/cockroach/pull/51437.

My concern is mainly that this distinction is not explicit right now and also that the semantics around generation 0 vs. generation non-0 are not spelled out anywhere, so I'm hesitant to begin relying on them.

andreimatei reviewed (commented):
So you'd want an explicit `speculative` flag in the `RangeCacheEntry`? Can do, but then I guess that forces our hand to have a way of resetting the `speculative` flag too when we know that the speculation has worked out. I guess this part becomes easy with #51437 when a successful RPC either returns an updated descriptor or implicitly confirms that the caller had it right. So I guess I can piggy-back an `EvictionToken.Confirm()` method on that PR - which would be a no-op for non-speculative descriptors and an eviction and reinsertion as non-speculative for the rest. Sounds like a plan?

Btw, @aayushshah15 demonstrated badness caused by the issue this PR is addressing using the kvnemesis with merges enabled. And I'm now pretty sure that this is also the cause of the assertion failure seen in #50995. So I guess I should write a test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MCiFzGzApRgnDq-UMfL:byy36d4)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

aayushshah15 commented:
> but it is the case that the speculative descriptors are required for splits to work correctly.
Lazy question, is this documented somewhere? If not, we should probably have a detailed section about this in the comments.





<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
Probably all we have is this comment: https://github.com/cockroachdb/cockroach/blob/a014e549b57f7e1ce33b82eb836cb466f93d15e4/pkg/kv/kvclient/kvcoord/dist_sender.go#L406-L410

Note to self: what I've discussed with Nathan is adding some type safety around speculative range cache entries by having the cache contain an interface as elements, with two implementations and a comparison function with a type switch.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MCiN0dYAssmTiK3UCJj:bmy4dh)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei committed:
physicalplan: narrow the information exposed by a RangeIterator

The replica oracles only need a leaseholder, not a full lease. This
patch a more narrow interface in the RangeIterator for exposing just
this instead of the fatter lease. This works towards not exposing leases
any more outside of the range cache since the range cache internally
deals with "speculative leases" hackily and we'd rather keep these
incomplete leases for escaping to clients.

Release note: None

andreimatei committed:
kvcoord: remove RangeIterator.Lease()

RangeIterator.Leaseholder() remains. Lease() had only one rando caller
left that doesn't matter much.
This patch is towards not exposing lease info from the
RangeDescriptorCache in order to not burden clients with details about
speculative vs non-speculative leases.

Release note: None

andreimatei committed:
roachpb: add a comment to NotLeaseholderError

Release note: None

andreimatei commented:
This is RFAL. I've given up on splitting speculative vs regular cache entries into different types because I became unconvinced it was helping everybody, but I've done other stuff to bring speculation more front and center.

nvb reviewed (commented):
The latest round is much better! Promoting speculative descriptors and leases to be front and center makes this all much easier to reason about.

> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 13 of 13 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MDVbxST3U5rQ2m5Bs1Y:b-z2u12w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvbase/range_cache.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVW7mCAc1NWkMY3heL:-MDVW7mCAc1NWkMY3heM:br24c20) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvbase/range_cache.go#L35)):*
> ```Go
> 	// DescSpeculative returns true if the descriptor in the entry is "speculative"
> 	// - i.e. it doesn't correspond to a committed value. Such descriptors have been
> 	// inserted in the cache with Generation=0.
> ```

Add a sentence describing how this is possible (i.e. that they were taken from intents).
___
*[pkg/kv/kvbase/range_cache.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVW--qEc4_1MeLPkSw:-MDVW--qEc4_1MeLPkSx:byu7naq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvbase/range_cache.go#L42)):*
> ```Go
> 	Leaseholder() *roachpb.ReplicaDescriptor
> 
> 	// Lease returns the cached lease, if known. Returns nil if no lease is known. It's
> ```

nit: rewrap some of these lines.
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1749 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVWMNw42a2FarplCFd:-MDVWMNw42a2FarplCFe:b-m8s5ac) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1749)):*
> ```Go
> 				replicas.MoveToFront(idx)
> 			} else {
> 				log.Eventf(ctx, "leaseholder %s missing from replicas", leaseholder)
> ```

Is this possible? Do we want it to be?
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1809 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVWewaAmqYxquc8mpZ:-MDVWewaAmqYxquc8mp_:b-22dts0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1809)):*
> ```Go
> 			// "speculative". Even if the speculation is correct, we want the serve to
> 			// return an update, at which point the cached entry will no longer be
> 			// "speculative".
> ```

nit: reference the `DescSpeculative()` method here.
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 226 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVauSs7BJbY3r9xopw:-MDVauSs7BJbY3r9xopx:bdyauki) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L226)):*
> ```Go
> 	// the opportunity to use this nextDesc; if another actor races to evict the
> 	// respective cache entry and wins, nextDesc becomes useless.
> 	nextDesc *roachpb.RangeDescriptor
> ```

Consider renaming this here and elsewhere to "speculativeDesc" and adding a bit of a comment describing its role.
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 314 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbMGj7Wv2lFHm0kHO:-MDVbMGj7Wv2lFHm0kHP:b60hvk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L314)):*
> ```Go
> //
> // It's legal to pass in a lease with a zero Sequence; it will be treated as a
> // speculative lease and considered newer than any existing lease.
> ```

But will be immediately replaced by any update, right?
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 836 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbTaZ11wzh9mf-TZm:-MDVbTaZ11wzh9mf-TZn:bvydt9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L836)):*
> ```Go
> 		if cachedEntry != nil && descsCompatible(cachedEntry.Desc(), entry.Desc()) {
> 			return false, cachedEntry
> 			//// cachedEntry is almost certainly newer than entry, but we still check
> ```

Did you mean to remove this?
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 1106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbc0T4ydGM6bmBro3:-MDVbc0T4ydGM6bmBro4:bcyh5ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L1106)):*
> ```Go
> // - i.e. it doesn't correspond to a committed lease. Such leases have been
> // inserted in the cache with Sequence=0.
> func (e *rangeCacheEntry) leaseSpeculative() bool {
> ```

Is it worth adding this to the interface as well? If only to show the symmetry to between DescSpeculative and LeaseSpeculative?
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 1125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVcPIu0g8a_m6D7GUp:-MDVcPIu0g8a_m6D7GUq:bmxlhq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L1125)):*
> ```Go
> // can't be determined what information is newer is when at least one of the
> // descriptors is "speculative" (generation=0), or when the lease information is
> // speculative.
> ```

nit: might as well give this word the same treatment: `"speculative (sequence=0)`.
___
*[pkg/roachpb/errors.go, line 424 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVUhw_4dYr8kydIujC:-MDVUhw_4dYr8kydIujD:b-mkd2np) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edbe4aeb3c7300e6690cb2222a8d5c01e920bf4/pkg/roachpb/errors.go#L424)):*
> ```Go
> 		return e.rangesInternal
> 	}
> 	// Fallback for 20.1 errors. Remove in 20.3.
> ```

Giving up hope?
___
*[pkg/roachpb/errors.proto, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVVGnN7ejPSYfIxeBd:-MDVVGnN7ejPSYfIxeBe:bbwpg0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a7d8bcfdfbf24d1fea24ad91a0951213aa2761b/pkg/roachpb/errors.proto#L41)):*
> ```protobuf
>   //
>   // It's possible for leases returned here to represent speculative leases, not
>   // actual committed leasea. In this case, the lease will not have its Sequence
> ```

s/leasea/leases/
___
*[pkg/roachpb/errors.proto, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVVN68AeCqeJLnvwef:-MDVVN68AeCqeJLnvweg:b-nuyboz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a7d8bcfdfbf24d1fea24ad91a0951213aa2761b/pkg/roachpb/errors.proto#L42)):*
> ```protobuf
>   // It's possible for leases returned here to represent speculative leases, not
>   // actual committed leasea. In this case, the lease will not have its Sequence
>   // set.
> ```

Should we codify this rule in a `(Lease) Speculative` method?

Also, do we need to worry about leases without sequence numbers on legacy clusters? I don't _think_ so because it seems like any active lease will have begun incrementing the lease sequence a few releases ago, but it's worth going through the thought experiment.
___
*[pkg/sql/physicalplan/replicaoracle/oracle.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVV1jiEM6QS0ubi1zT:-MDVV1jiEM6QS0ubi1zU:b-hm48w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb939e2ee78832c768eeae83088ce7d28419b3e2/pkg/sql/physicalplan/replicaoracle/oracle.go#L71)):*
> ```Go
> 	// about any of the nodes that might be tried.
> 	ChoosePreferredReplica(
> 		ctx context.Context, rng *roachpb.RangeDescriptor, leaseHolder *roachpb.ReplicaDescriptor, qState QueryState,
> ```

nit: is leaseholder one word or two? I don't have strong preferences, but let's be consistent. What do we do elsewhere?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvbase/range_cache.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVW7mCAc1NWkMY3heL:-MDuWqMY642WWrYV3vqP:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvbase/range_cache.go#L35)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a sentence describing how this is possible (i.e. that they were taken from intents).
</blockquote></details>

done
___
*[pkg/kv/kvbase/range_cache.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVW--qEc4_1MeLPkSw:-MDuWte58uB__veMcYAG:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvbase/range_cache.go#L42)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: rewrap some of these lines.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1749 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVWMNw42a2FarplCFd:-MDuYdi04fkTBGZ_Ud6f:b-q7wtt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1749)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this possible? Do we want it to be?
</blockquote></details>

I think so. Added a comment explaining how it can happen.
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1809 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVWewaAmqYxquc8mpZ:-MDuYzOTDQCRCtQt2QV8:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1809)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: reference the `DescSpeculative()` method here.
</blockquote></details>

done
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 226 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVauSs7BJbY3r9xopw:-MDuZzMo8RTZCZGy60pU:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L226)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider renaming this here and elsewhere to "speculativeDesc" and adding a bit of a comment describing its role.
</blockquote></details>

see now
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 314 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbMGj7Wv2lFHm0kHO:-MDu_H48DPUlCBiTEi-e:b-cbugmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L314)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

But will be immediately replaced by any update, right?
</blockquote></details>

expanded the comment
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 836 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbTaZ11wzh9mf-TZm:-MDu_SXo0GIfVI9u35Sa:blrpli9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L836)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you mean to remove this?
</blockquote></details>

yeah, thanks
I've removed it cause I thought handling  it's not worth the code (and it was giving me some other trouble too I think)
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 1106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVbc0T4ydGM6bmBro3:-MDuaFgG1p2rjXlzxm9z:b-79vlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L1106)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it worth adding this to the interface as well? If only to show the symmetry to between DescSpeculative and LeaseSpeculative?
</blockquote></details>

done

but btw I'm gonna move the range cache to its own package and get rid of the `RangeCacheEntry` iface in `kvbase`
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 1125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVcPIu0g8a_m6D7GUp:-MDuaWw973Xxa6gjTh-h:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b06660d666b27c1e0ae64f64ea6d4f82fb32f04e/pkg/kv/kvclient/kvcoord/range_cache.go#L1125)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: might as well give this word the same treatment: `"speculative (sequence=0)`.
</blockquote></details>

done
___
*[pkg/roachpb/errors.go, line 424 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVUhw_4dYr8kydIujC:-MDufbQwC2Rz9xPAHPwq:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edbe4aeb3c7300e6690cb2222a8d5c01e920bf4/pkg/roachpb/errors.go#L424)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Giving up hope?
</blockquote></details>

:)
___
*[pkg/roachpb/errors.proto, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVVGnN7ejPSYfIxeBd:-MDuy7-569sM3riDiEgy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a7d8bcfdfbf24d1fea24ad91a0951213aa2761b/pkg/roachpb/errors.proto#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/leasea/leases/
</blockquote></details>

done
___
*[pkg/roachpb/errors.proto, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVVN68AeCqeJLnvwef:-MDuwRx39USGxMVld6u0:b-vl1rvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a7d8bcfdfbf24d1fea24ad91a0951213aa2761b/pkg/roachpb/errors.proto#L42)):*
> Should we codify this rule in a (Lease) Speculative method?

done, and used the new method in the next commit

> Also, do we need to worry about leases without sequence numbers on legacy clusters? I don't think so because it seems like any active lease will have begun incrementing the lease sequence a few releases ago, but it's worth going through the thought experiment.

yeah, I've thought about this and I don't think we need to worry about it. One thing that I've tested is that a new cluster doesn't produce leases with seq=0 even immediately after bootstrapping, and it doesn't.
___
*[pkg/sql/physicalplan/replicaoracle/oracle.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVV1jiEM6QS0ubi1zT:-MDuzKNc98qtpsf1XXuh:b-cv1r4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb939e2ee78832c768eeae83088ce7d28419b3e2/pkg/sql/physicalplan/replicaoracle/oracle.go#L71)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: is leaseholder one word or two? I don't have strong preferences, but let's be consistent. What do we do elsewhere?
</blockquote></details>

one word is more prevalent (at least when there's not other words in the variable name). Switched to that.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 3 of 18 files at r5, 2 of 3 files at r7, 13 of 13 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MENujvR3YdzYx7aaZq1:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/roachpb/errors.proto, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-MDVVN68AeCqeJLnvwef:-MENuZNDDKzjMIUIAlRQ:bmj11fc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a7d8bcfdfbf24d1fea24ad91a0951213aa2761b/pkg/roachpb/errors.proto#L42)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> Should we codify this rule in a (Lease) Speculative method?

done, and used the new method in the next commit

> Also, do we need to worry about leases without sequence numbers on legacy clusters? I don't think so because it seems like any active lease will have begun incrementing the lease sequence a few releases ago, but it's worth going through the thought experiment.

yeah, I've thought about this and I don't think we need to worry about it. One thing that I've tested is that a new cluster doesn't produce leases with seq=0 even immediately after bootstrapping, and it doesn't.
</blockquote></details>

Thanks for running that test. The moment after bootstrapping was also concerning me.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51395#-:-MEOVFrEBmECio0j9sJc:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
